028-service-binding.md

```diff
@@ -0,0 +1,374 @@
+
+Required changes:
+
+ - Add a new Custom Resource called `KafkaConnection` that refers to the `Kafka`, the specific listener and the `KafkaUser` that should be used
+ - The Strimzi operator adds a `status.binding` to the `KafkaConnection` status that points to a Kubernetes `Secret`
+ - The Strimzi operator creates a `Secret` in the same namespace as the `KafkaConnection` containing all the required information
+
+`KafkaConnection` CR spec:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnection
+metadata:
+  name: barista-kafka
+spec:
+  kafka:
+    apiVersion: kafka.strimzi.io/v1beta2
+    kind: Kafka
+    name: my-cluster
+    namespace: my-namespace # optional
+    listener: # optional
+      name: tls # optional
+  user: # optional
+    apiVersion: kafka.strimzi.io/v1beta2
+    kind: KafkaUser
+    name: my-barista
+    namespace: my-namespace # optional
+status:
+  binding:
+    name: barista-kafka
+```
+
+`Secret` created by Strimzi:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: barista-kafka
+  namespace: my-namespace
+type: Opaque
+  data:
+    type: kafka
+    provider: strimzi
+    bootstrapServer: # comma separated list of host:port
+  # Provided if TLS enabled:
+    ca.crt: #  Strimzi cluster CA certificate
+    ca.p12: #  PKCS #12 archive file for Strimzi cluster CA certificate
+    ca.password: # Password for protecting the Strimzi cluster CA certificate PKCS #12 archive file
+  # Provided if selected user is SCRAM auth:
+    username: # SCRAM username
+    password: # SCRAM password
+    sasl.jaas.config: # sasl jaas config string for use by Java applications
+  # Provided if selected user is mTLS:
+    user.p12: # client certificate for the consuming client PKCS #12 archive file for storing certificates and keys
+    user.password: # password for protecting the client certificate PKCS #12 archive file
+    user.crt: # certificate for the consuming client signed by the clients' CA
+    user.key: # private key for the consuming client
+```
+
+Naming of secret keys:
+
+The names for the keys given above match what is currently provided in other secrets created by Strimzi. The service binding specification does included some suggested fields for certificates, but they do not satisfy the requirements of an application connecting to Strimzi. These fields could be proposed back to the Service Binding specification.
+
+An alternative approach for naming is to name the fields to match the use, for example `ca` -> `truststore` (`truststore.crt`, `truststore.p12`, `truststore.password`), `user` -> `keystore` (`keystore.p12`, `keystore.password`, `keystore.crt`, `keystore.key`)
+ 
+
+### Implementation
+
+This proposal will result in a new `KafkaConnectionAssemblyOperator` class being added to the `cluster-operator` directory in the `strimzi-kafka-operator` repo. The class will be 
```

Comment:
It's a total aside, but I'm fed up with the confusing class naming we've ended up with. `KafkaConnectionController` (and renaming the other `*AssemblyOperator` to be `*Controller` and the various "resource operators" to be `*Operations` would help us when talking about our code, and likely be more understandable for people new to the project). Wdyt?

Reply:
I agree that the names are confusing, I don't know what `AssemblyOperator` actually means. I think `Controller` would make it more obvious what the purpose of the class is.

Reply:
WDYT @scholzj @ppatierno?

Reply:
I would argus that our existing `*AssemblyOperator` classes are not controllers. They do not follow the Kubernetes controller pattern. So I think renaming them to controller is confusing. TBH, I think the `Assembly` part is much more confusing part of the name than the `Operator` part.

Reply:
AFAIK when you have a custom resource to handle you are dealing with an "operator". I agree with Jakub that the `Assembly` part is the most confusing one, I have never seen this as part of operators developed with tools like the Go sdk for example. I think we should stay with `Operator` and trying at some point to remove that `Assembly` part. Maybe we can start with this new operator to do that.

Reply:
Having `KafkaOperator` and `KafkaConnectOperator` etc would be an improvement, but there's still the confusion with the "Cluster Operator" and "Entity Operator" terms, which are used extensively in the docs. Maybe we just can't avoid that confusion.

But what about the "resource operator" classes? Those are not really operators (they have no watch). At best they're "reconcilers", but we use methods like `get`/`getAsync` as much as `reconcile`. Having a better term for them would also be an improvement, simply from having fewer `*Operator` classes floating around.

Reply:
For what it's worth it looks like the java-operator-sdk project have currently landed on "Reconciler" as the name for the class that is responsible for updating a resource to look how it should.

Reply:
> But what about the "resource operator" classes? Those are not really operators (they have no watch). At best they're "reconcilers", but we use methods like get/getAsync as much as reconcile. Having a better term for them would also be an improvement, simply from having fewer *Operator classes floating around.

To be honest, these are really just `*Utils` classes, right? They have a `reconcile` method, but they do not run the reconciliation loop. But TBH, I'm not sure the current naming is really a blocker for anything and being consistent at this point is more useful than changing the names around in a hope they will be more clear to someone else.